packages/arb-validator-core/gobridge/goEth.go

```diff
@@ -0,0 +1,370 @@
+		mEthData.nextAddress = common.BigIntToAddress(big.NewInt(1))
+		mEthData.outchans = make(map[chan arbbridge.MaybeEvent]void)
+		mEthData.chanMgr = make(chan chan arbbridge.MaybeEvent)
+		mEthData.pubchan = make(chan arbbridge.MaybeEvent)
+		mEthData.ethWallet = make(map[common.Address]*big.Int)
+		//mEthData.ChannelWallet = make(map[common.Address]protocol.TokenTracker)
+		//mEthData
+		//onceMutex.Unlock()
+		go func() {
+			for {
+				select {
+				case ch := <-mEthData.chanMgr: // register outchan
+					mEthData.outchans[ch] = Void
+				case msg := <-mEthData.pubchan: // publish to outchans
+					for ch := range mEthData.outchans {
+						ch <- msg
+					}
+				}
+			}
+		}()
+		go func() {
+			for x := range time.Tick(2 * time.Second) {
+				mine(mEthData, x)
+			}
+		}()
+	})
+	return GoEth[ethURL]
+}
+
+func (m *goEthdata) getNextAddress() common.Address {
+	addr := m.nextAddress
+	addrInt := new(big.Int)
+	addrInt.SetBytes(addr[:])
+	m.nextAddress = common.BigIntToAddress(addrInt.Add(addrInt, big.NewInt(1)))
+	return addr
+}
+
+func (m *goEthdata) getCurrentBlock() *common.BlockId {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+
+	return m.NextBlock
+}
+
+func (m *goEthdata) getLastBlock() *common.BlockId {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+
+	return m.LastMinedBlock
+}
+
+func (m *goEthdata) getBlockFromHeight(height *common.TimeBlocks) (*common.BlockId, error) {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+
+	b, ok := m.blockNumbers[height.AsInt().Uint64()]
+	if !ok {
+		return nil, errors.New("not found")
+	}
+
+	return b, nil
+}
+
+func (m *goEthdata) registerOutChan(oc chan arbbridge.MaybeEvent) {
+	m.chanMgr <- oc
+}
+
+func (m *goEthdata) pubMsg(challenge *challengeData, msg arbbridge.MaybeEvent) {
+	m.msgMutex.Lock()
+	defer m.msgMutex.Unlock()
+	m.nextMsgs[msg.Event.GetChainInfo().BlockId] = append(m.nextMsgs[msg.Event.GetChainInfo().BlockId], goMaybeEvent{
+		challenge: challenge,
+		event:     msg,
+	})
+}
+
+func mine(m *goEthdata, t time.Time) {
+	m.blockMutex.Lock()
+	blockHash := hashing.SoliditySHA3(hashing.Uint256(big.NewInt(rand2.Int63())))
+	m.NextBlock.HeaderHash = blockHash
+	lastBlock := m.LastMinedBlock
+	m.blockNumbers[m.NextBlock.Height.AsInt().Uint64()] = m.NextBlock
+	m.blockHashes[m.NextBlock.HeaderHash] = m.NextBlock
+	m.parentHashes[*m.NextBlock] = lastBlock.HeaderHash
+	m.LastMinedBlock = m.NextBlock
+	m.msgMutex.Lock()
+	for _, event := range m.nextMsgs[m.NextBlock] {
+		for _, ru := range m.rollups {
+			ru.events[m.NextBlock] = append(ru.events[m.NextBlock], event.event.Event)
+		}
+		m.challengeWatchersMutex.Lock()
+		for challenge, watcher := range m.challengeWatcherEvents {
+			// if this is a challenge specific event and it is not for this challenge ignore it
+			if event.challenge != nil && event.challenge != challenge {
+				continue
+			}
+			watcher[m.NextBlock] = append(watcher[m.NextBlock], event.event.Event)
+		}
```

Comment:
Why not just key event address to contract rather than doing these loops?

Reply:
cleaned up event handling